rfc-161-technical-direction-in-publishing-2023.md

```diff
@@ -0,0 +1,144 @@
+- Is this new software need something the GOV.UK programme will continue investing in for the foreseeable future? If not, it’s unwise to build it.
+- Is this a need for a specific department? If so, can we abstract the specific need to a common need to avoid the ownership confusion of us building a tool for one department
+
+#### Embrace existing conventions; start new conventions collaboratively
+
+GOV.UK software, like most software, is built upon layers of convention and standards. We like conventions: they reduce developer surprises and unnecessary decisions. We don’t like competing conventions - we prefer one consistent, if non-ideal, approach to solve a problem applied to an app rather than individual developers picking their individual preference each time they face a similar problem.
+
+Many GOV.UK conventions are documented. There is the [Service Manual](https://www.gov.uk/service-manual) on the high level team approach. There are organisational tech standards in the [GDS Way](https://gds-way.cloudapps.digital/). GOV.UK has [conventions for Rails applications](https://docs.publishing.service.gov.uk/manual/conventions-for-rails-applications.html) and individual repositories can have [team agreed conventions](https://github.com/alphagov/whitehall/blob/main/docs/adr/0001-add-view-components.md). 
+
+Not all conventions are documented. Some are learned from reading the code and the early decisions the developers made; others can be learned from observing team members or through code reviews. Not all conventions are in how software is written or documented. Others are in how team members interact with each other and collaborate.
+
+Starting new conventions is a team activity not an individual one - we want the results of our teams to feel collective. Please communicate liberally in starting a convention so it has the best chance of becoming established and adopted.
+
+#### Code should be correct, clear, concise and tested
+
+Code should be optimised for readability. Choose clarity over brevity or performance, unless there is a genuine need not to. 
+
+Code should have automated tests. These tests should be [necessary and sufficient](https://blog.testdouble.com/posts/2020-02-25-necessary-and-sufficient/).
+
+Embrace relevant software engineering principles such as [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns), [single responsibily principle](https://en.wikipedia.org/wiki/Single-responsibility_principle), [You aren’t gonna need it (YAGNI)](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it) and [The boy scout rule](https://www.oreilly.com/library/view/97-things-every/9780596809515/ch08.html). Embrace consistency, modularity and abstractions to manage complexity. Place value in [intuitive naming](https://medium.com/signal-v-noise/hunting-for-great-names-in-programming-16f624c8fc03).
```

Comment:
Typo: "responsibily".

Speaking of intuitive naming: the name "the boy scout rule" is not particularly intuitive. Could we come up with something more intuitive and more inclusive? ("Drive-by improvements"? "Opportunistic enhancements"?)

Reply:
Thanks re the typo, I've fixed it.

Yeah, I was a bit uncomfortable with the gendering of Boy Scout rule and did spend some time searching for a linkable principle that was effectively a synonym and didn't find a good example. Other suggestions would be great. We could potentially just replace it as a linked principle with a sentence of "Value that when changing code, you leave it in as-good or better state than in how you found it".

Reply:
I think I've heard people just talk about it as the Campsite Rule:
- https://wolfgang-ziegler.com/blog/Development-Practice-The-Campsite-Rule.aspx 
- https://medium.com/@fooyay/the-campsite-rule-60cbf1054d92
- https://cultivatingsoftware.wordpress.com/2013/09/09/the-campsite-rule/

...which is ungendered although only slightly more intuitive.

(although: I would say maybe this is an okay use of the term because we're using it as a link text to a page which is headed "The Boy Scout Rule"? I wouldn't say "single responsibility principle" or "you aren't gonna need it" are super intuitive names either, so the fact you've linked to better explanations sort of covers the fact that naming is hard).

Reply:
The Campsite Rule reads better! And it allows us to link the definition to anyone other than Robert C Martin, so we don't endorse someone[1] with a long[2] history[3] of exclusionary[4] behaviour[5] in a document giving technical direction to the entire Publishing community.

[1] https://blog.wesleyac.com/posts/robert-martin
[2] https://corgibytes.com/blog/2019/02/14/why-we-changed-a-core-value/
[3] https://www.madetech.com/blog/no-uncle-of-mine/
[4] https://web.archive.org/web/20200713003355/https://techexplained.substack.com/p/tech-bullshit-explained-uncle-bob
[5] https://web.archive.org/web/20200914214901mp_/https://meaganwaller.com/framework-whipped/

Reply:
Thanks, yup also good shout on avoiding reference to a divisive figure.

I've gone with adding the words directly rather than referencing "the Campsite Rule" as I felt that, unlike the other terms, there wasn't likely to be developer familiarity. By googling it, I also found that the phrase "Campsite Rule" has some sexual relationship usage - albeit in a very similar context to this - that made me a little wary.

Change is in: https://github.com/alphagov/govuk-rfcs/compare/75a520efa36f5caf802c6e061efbca0eb164ddcd..9c2af0c500d17ff9bcea2b3e8064442d182f04ae